add CSV docs (#62)
xataio/mdx-docs: 080-Recipes/010-csv-data/010-import-csv-data-UI.mdx

---
title: Import CSV data
navTitle: Import CSV data
keywords: ['csv', 'import', 'migrate data']
description: Import records from a local CSV file into a Xata database table
slug: recipes/import-data
published: true
---

## Overview

A CSV file is a standard text file where data fields are typically divided by commas (,), and saved with the `.csv` extension.
You can import and work with CSV file data in your Xata database.

> **NOTE:** While CSV files are explicity mentioned, Xata supports the import of various file formats, including tab-separated values (TSV), comma-separated values (CSV), delimiter-separated values (DSV), and plain text data. This covers a range of extensions including `.tsv`, `.csv`, `.dsv`, and `.txt` formats.

Reviewer: This might be worth mentioning somewhere, but I'm wondering if it's a bit too prominent at the moment?

You can import CSVs using the following methods:

> - Import a CSV file using the Xata UI (recommeneded)

Reviewer: ```suggestion
- Import a CSV file using the Xata UI (recommended)
```

- Import a CSV file using the Xata CLI

### Considerations

> - Decide if you are using the Xata UI (Workspace) or CLI to import CSV files. Using the UI is the recommended option, but you can use both to edit records after importing.

Reviewer: ```suggestion
- Decide if you are using the Xata UI or CLI to import CSV files. Using the UI is the recommended option, but you can use both to edit records after importing.
```

> - Ensure your file's accuracy before importing. Certain values may become uneditable after importing. For errors in the imported file, you'll need to delete, edit the CSV file, and try importing again.

Reviewer: > Certain values may become uneditable after importing

The only thing this is true for is ids for most values you can edit them in the UI

> - Ensure your file's accuracy before importing. Certain values may become uneditable after importing. For errors in the imported file, you'll need to delete, edit the CSV file, and try importing again.

Reviewer: ```suggestion
- Please check your file before importing. It's best to resolve any data issues before uploading the file, you can also edit your data in the UI after import. 
```

Low commitment level!

| ------------- | --------------------------------------------------------- | --------------------------------- |
| String        | Basic textual information                                 | "Value"                           |
| Text          | Longer pieces of textual content                          | "Hello world and welcome to Xata" |
| Integer       | Whole numbers without decimal points                      | `42`                              |
| Float         | Numbers with decimal points                               | `3.14`                            |
| Boolean       | Represents true or false values                           | `true`                            |
| Datetime      | Date and time information                                 | 2023-08-30T00:00:00               |
| Link to Table | Creates relationships between tables                      | Table: Orders                     |
| Email         | Email addresses                                           | `user@example.com`                |
| Multiple      | Multiple values from a predefined set separated by commas | \["Red", "Green"]                 |

> | File          | Reference to an external file, attachment, or document    | `file.pdf`                        |

Reviewer: Not sure about the description column, should / could we link our core docs for each column type (which might also need updating 🧶 ):  /docs/concepts/data-model#string etc.

We don't support file or vector. My vote is to include them in the table and say: "Not supported" to make it clear.

> | Format Type   | Description                                               | Example                           |
> | ------------- | --------------------------------------------------------- | --------------------------------- |
> | String        | Basic textual information                                 | "Value"                           |
> | Text          | Longer pieces of textual content                          | "Hello world and welcome to Xata" |
> | Integer       | Whole numbers without decimal points                      | `42`                              |
> | Float         | Numbers with decimal points                               | `3.14`                            |
> | Boolean       | Represents true or false values                           | `true`                            |
> | Datetime      | Date and time information                                 | 2023-08-30T00:00:00               |
> | Link to Table | Creates relationships between tables                      | Table: Orders                     |
> | Email         | Email addresses                                           | `user@example.com`                |
> | Multiple      | Multiple values from a predefined set separated by commas | \["Red", "Green"]                 |
> | File          | Reference to an external file, attachment, or document    | `file.pdf`                        |

Reviewer: ```suggestion
| Format Type   | Description                                               | Example formats                           |
| ------------- | --------------------------------------------------------- | --------------------------------- |
| String        | With or without quotes, less than 180 characters.                                 | `"Value"` or `Value`                           |
| Text          |  With or without quotes, longer than 180 characters.                       | `"Hello world and welcome to Xata" |
| Integer       | Whole numbers without decimal points                      | `42`                              |
| Float         | Numbers with decimal points                               | `3.14`                            |
| Boolean       | Represents true or false values                           | `true`                            |
| Datetime      | Our importer will attempt to parse all common datetimes                          | `2023-08-30T00:00:00.000Z` or `2023-08-30T00:00:00.000Z` or `26-02-88` or any other date format               |
| Link to Table | Id of a record in another table                 | `rec_xyz` or `usr_xyz` if using custom ids                     |
| Email         | Email addresses                                           | `user@example.com`                |
| Multiple      | Comma seperated or a JSON array | `Red,Green` or `["Red", "Green"]`                 |
| File          | Not supported    |                         |
| Vector          | Not supported    |                         |
```

### Filename formatting

- Ensure that the imported filename/table name is limited to 64 characters or fewer.
- The table name should consist of alphanumeric characters, hyphens (-), and underscores(\_). Special characters are not allowed.

### Delimiter

A delimiter acts as dividers and distinguish individual data fields within a CSV file. They are useful in structuring the data, allowing for accurate interpretation and parsing.

- The comma (,) is applied as the default delimiter for the CSV import function.
- The CSV import functionality supports common delimiters such as commas (,), semicolons (;), quotes ( ", ' ), braces ({}), pipes (|), or slashes ( / \ ). You can modify the delimiter in the UI, when importing the CSV file data.

## Import a CSV file using the UI

> Xata offers an import tool directly within your dashboard Workspace. This tool enables seamless importing of existing files into the Xata database without causing any downtime. As an alternative, you also have the option to import CSV data using the CLI; however, it's worth noting that the UI method is the recommended approach.
> Importing the CSV file involves transfering data as well as establishing a schema. Once imported, the data in the CSV file will be accessible as a table in your workspace.

Reviewer: ```suggestion
Xata offers an import tool directly within your dashboard Workspace. 
```

1.  Log in to your Xata account and go to the desired Workspace and database. If you haven't created a Workspace as yet, click **Add a Workspace** to do so.

> 2.  Choose the preferred branch for your work.

Reviewer: For many people choosing the branch might mean no action. Even though it's slightly misleading I think we could delete step 2 entirely for the greater good.

3.  On the left-hand menu under "Tables," click **Add a table** > **Import a CSV file**.
4.  Select **Browse Desktop** to choose a supported file to import. Ensure the filename adheres to the formatting recommendations.
5.  Click **Begin import** to initiate the import.

> You'll have acces to partially imported records as the process completes.

Reviewer: I'd delete this line, we have a very verbosely name button

> 

Reviewer: Something very important, that is possibly worth noting with like a ⚠️ is that they need to leave the tab open until the import completes.

---
title: Import CSV data
navTitle: Import CSV data
keywords: ['csv', 'import', 'migrate data']
description: Import records from a local CSV file into a Xata database table

> slug: csv-data/import-data-ui

Reviewer: ```suggestion
slug: csv-data/import-data```

published: true
---

## Overview

A CSV file is a standard text file where data fields are typically divided by commas (,), and saved with the `.csv` extension.
You can import and work with CSV file data in your Xata database.

**NOTE:** While CSV files are explicity mentioned, Xata supports the import of various file formats, including tab-separated values (TSV), comma-separated values (CSV), delimiter-separated values (DSV), and plain text data. This covers a range of extensions including `.tsv`, `.csv`, `.dsv`, and `.txt` formats.

You can import CSVs using the following methods:

> - [Import a CSV file using the Xata UI](/docs/recipes/csv-data/import-data-ui#import-a-csv-file-using-the-ui) (recommeneded)

Reviewer: ```suggestion
- [Import a CSV file using the Xata UI](/docs/recipes/csv-data/import-data-ui#import-a-csv-file-using-the-ui) (recommended)
```

- Ensure your file's accuracy before importing. Certain values may become uneditable after importing. For errors in the imported file, you'll need to delete the table, edit the CSV file, and try importing again.

## Formatting data

When importing CSV files into Xata, ensure the data aligns with the correct format for the database. You can add additional columns and specify data types after importing data.

### Format types

Choose a format that matches the data you're entering. Xata columns support the following formats:

| Format Type   | Description                                                                            | Example formats                                                |
| ------------- | -------------------------------------------------------------------------------------- | -------------------------------------------------------------- |
| String        | Text with or without quotes, less than 180 characters.                                 | `"Value"` or `Value`                                           |

> | Text          | Text with or without quotes, longer than 180 characters.                               | `"Hello world and welcome to Xata"                             |

Reviewer: ```suggestion
| Text          | Text with or without quotes, longer than 180 characters.                               | `"Hello world and welcome to Xata"`                             |
```

| Integer       | Whole numbers without decimal points                                                   | `42`                                                           |
| Float         | Numbers with decimal points                                                            | `3.14`                                                         |
| Boolean       | Represents true or false values                                                        | `true`                                                         |

> | Datetime      | Date and time information. Our CSV importer will attempt to parse all common datetimes | `2023-08-30T00:00:00.000Z`or`26-02-88`or any other date format |

Reviewer: ```suggestion
| Datetime      | Date and time information. Our CSV importer will attempt to parse all common datetimes | `2023-08-30T00:00:00.000Z` or `26-02-88` or any other date format |
```

> | Link to Table | ID of a record in another table                                                        | `rec\_xyz`or`usr\_xyz`if using custom IDs                      |

Reviewer: ```suggestion
| Link to Table | ID of a record in another table                                                        | `rec_xyz` or `usr_xyz`if using custom IDs                      |
```

| Email         | Email addresses                                                                        | `user@example.com`                                             |

> | Multiple      | Multiple values seperated by a comma or a JSON array                                   | `Red,Green`or`\["Red", "Green"]\`                              |

Reviewer: ```suggestion
| Multiple      | Multiple values seperated by a comma or a JSON array                                   | `Red,Green`or`["Red", "Green"]`                              |
```

| File          | Not supported                                                                          |                                                                |
| Vector        | Not supported                                                                          |                                                                |

### Filename formatting

- Ensure that the imported filename/table name is limited to 64 characters or fewer.

> - The table name should consist of alphanumeric characters, hyphens (-), and underscores(\_). Special characters are not allowed.

Reviewer: That's fine - no problem

### Delimiter

> A delimiter acts as dividers and distinguish individual data fields within a CSV file. They are useful in structuring the data, allowing for accurate interpretation and parsing.

Reviewer: ```suggestion
A delimiter acts as dividers and distinguishes individual data fields within a CSV file. They are useful in structuring the data, allowing for accurate interpretation and parsing.
```

- The CSV import function has the capability to automatically identify the delimiter used in your CSV file. In case your file follows an uncommon format, you have the option to adjust the delimiter using the Xata UI.

> - The CSV import functionality supports common delimiters such as commas (,), semicolons (;), quotes ( ", ' ), braces ({}), pipes (|), or slashes ( / \ ). You can modify the delimiter in the UI, when importing the CSV file data.

Reviewer: ```suggestion
- The CSV import functionality supports common delimiters such as commas `,`, semicolons `;`, quotes `"`, `'`, braces `{`, `}`, pipes `|`, or slashes `/`, `\`. You can modify the delimiter in the UI, when importing the CSV file data.
```

## Import a CSV file using the UI

Xata offers an import tool directly within your dashboard Workspace. This tool allows you to import exist files into the Xata database without causing any downtime. As an alternative, you also have the option to import CSV data using the CLI; however, it's worth noting that the UI method is the recommended approach.

> Importing the CSV file involves transfering data as well as establishing a schema. Once imported, the data in the CSV file will be accessible as a table in your database Workspace.

Reviewer: ```suggestion
Importing the CSV file involves establishing a schema and then transferring data. Once imported, the data in the CSV file will be accessible as a table in your database Workspace.
```

1.  Log in to your Xata account and go to the desired Workspace and database. If you haven't created a Workspace as yet, click **Add a Workspace** to do so.
2.  Choose the preferred branch for your work.
3.  The left-hand menu under "Tables," click **Add a table** > **Import a CSV file**.
    ![Import CSV files](../images/import-csv-file.png)
4.  Select **Browse Desktop** to choose a supported file to import. Ensure the filename adheres to the formatting recommendations.

> 5.  Click **Begin import** to initiate the import. You'll have access to partially imported records as the process completes.

Reviewer: I think there is a step missing between 4 & 5 where they:

- Select column types (we should highlight these cannot be changed after import)
- Rename columns

<Alert status="warning">
  To ensure all data from the chosen CSV file is imported successfully, keep the tab open until the import process
  finishes.
</Alert>

> ## View imported CSV table
> 
> After importing CSV data into Xata, you can use sorting and filtering capabilities to enhance data organization and retrieval from your table.
> 
> Xata allows you to sort your imported data based on specific columns. This can help you arrange information in ascending or descending order, making it easier to identify trends, patterns, or specific values within your dataset.
> To sort data after importing your CSV file, navigate to the relevant table and select the column header you want to sort by. Xata will arrange the data based on your choice.
> 
> Filtering allows you to narrow down the data you're working with. To filter data, access the table containing your imported CSV data and select the filter options available for each column and format. You can apply filters based on conditions or values that you want to isolate.

Reviewer: ```suggestion
```

This isn't really related to CSV import at all? If we want this content I think it should live in a different page in a seperate PR.

### Edit table

Each row within the imported table is editable. To edit row values in your table:

1.  Select the row you want to modify.
2.  Click the edit icon at the end of the row.

Note that certain system-generated attributes such as `xata.createdAt`, `xata.updatedAt`, and `xata.version` cannot be edited.

#### Column type consistency

> Once you define the column type during the initial configuration, it cannot be changed afterward. This ensures data consistency and accuracy within your table.

Reviewer: I've suggested mentioning this higher up, if you're following our instructions it's too late if you read it here. If we do mention this higher up - we should delete it here.
